src/world/map.c

```diff
@@ -892,6 +1293,335 @@ void game_command_change_surface_style(int* eax, int* ebx, int* ecx, int* edx, i
+{
+	int ax = (uint16)*eax;
+	int ay = (uint16)*ecx;
+	int bx = (uint16)*edi;
+	int by = (uint16)*ebp;
+
+	int cost = 0;
+
+	uint8 max_height = 0xFF;
+
+	for(int yi = ay; yi <= by; yi += 32){
+		for(int xi = ax; xi <= bx; xi += 32){
+			rct_map_element* map_element = map_get_surface_element_at(xi / 32, yi / 32);
+			uint8 height = map_element->base_height;
+			if(map_element->properties.surface.terrain & 0x1F){
+				height = (map_element->properties.surface.terrain & 0x1F) * 2;
+			}
+			if(max_height > height){
+				max_height = height;
+			}
+		}
+	}
+
+	for(int yi = ay; yi <= by; yi += 32){
+		for(int xi = ax; xi <= bx; xi += 32){
+			rct_map_element* map_element = map_get_surface_element_at(xi / 32, yi / 32);
+			
+			if(map_element->base_height <= max_height){
+				uint8 height = (map_element->properties.surface.terrain & 0x1F);
+				if(height){
+					height *= 2;
+					if(height > max_height){
+						continue;
+					}
+					height += 2;
+				}else{
+					height = map_element->base_height + 2;
+				}
+				int eax2 = xi, ebx2 = *ebx, ecx2 = yi, edx2 = (max_height << 8) + height, esi2, edi2, ebp2;
+				ebx2 = game_do_command_p(GAME_COMMAND_16, &eax2, &ebx2, &ecx2, &edx2, &esi2, &edi2, &ebp2);
+				if(ebx2 == MONEY32_UNDEFINED){
+					*ebx = MONEY32_UNDEFINED;
+					return;
+				}else{
+					cost += ebx2;
+				}
+			}
+		}
+	}
+	if(*ebx & GAME_COMMAND_FLAG_APPLY){
+		int x = ((ax + bx) / 2) + 16;
+		int y = ((ay + by) / 2) + 16;
+		int z = map_element_height(x, y);
+		sint16 water_height_z = z >> 16;
+		sint16 base_height_z = z;
+		z = water_height_z;
+		if(!z){
+			z = base_height_z;
+		}
+		RCT2_GLOBAL(0x009DEA5E, uint32) = x;
+		RCT2_GLOBAL(0x009DEA60, uint32) = y;
+		RCT2_GLOBAL(0x009DEA62, uint32) = z;
+		sound_play_panned(SOUND_LAYING_OUT_WATER, 0x8001, x, y, z);
+	}
+	*ebx = cost;
+}
+
+/**
+ *
+ *  rct2: 0x006E6878
+ */
+void game_command_lower_water(int* eax, int* ebx, int* ecx, int* edx, int* esi, int* edi, int* ebp)
+{
+	int ax = (uint16)*eax;
+	int ay = (uint16)*ecx;
+	int bx = (uint16)*edi;
+	int by = (uint16)*ebp;
+
+	int cost = 0;
+
+	uint8 min_height = 0;
+
+	for(int yi = ay; yi <= by; yi += 32){
+		for(int xi = ax; xi <= bx; xi += 32){
+			rct_map_element* map_element = map_get_surface_element_at(xi / 32, yi / 32);
+
+			uint8 height = map_element->properties.surface.terrain & 0x1F;
+			if(height){
+				height *= 2;
+				if(height > min_height){
+					min_height = height;
+				}
+			}
+		}
+	}
+
+	for(int yi = ay; yi <= by; yi += 32){
+		for(int xi = ax; xi <= bx; xi += 32){
+			rct_map_element* map_element = map_get_surface_element_at(xi / 32, yi / 32);
+			
+			uint8 height = (map_element->properties.surface.terrain & 0x1F);
+			if(height){
+				height *= 2;
+				if(height < min_height){
+					continue;
+				}
+				height -= 2;
+				int eax2 = xi, ebx2 = *ebx, ecx2 = yi, edx2 = (min_height << 8) + height, esi2, edi2, ebp2;
+				ebx2 = game_do_command_p(GAME_COMMAND_16, &eax2, &ebx2, &ecx2, &edx2, &esi2, &edi2, &ebp2);
```

Comment:
i don't think you have to use the p version of game_do_command here as you only require the cost returned. It saves on the requirement for defining all those extra variables.